closed/adds/jdk/src/share/classes/sun/security/ec/NativeECDHKeyAgreement.java

```diff
@@ -0,0 +1,264 @@
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+
+import jdk.crypto.jniprovider.NativeCrypto;
+
+import sun.security.action.GetPropertyAction;
+import sun.security.util.NamedCurve;
+
+/**
+ * Native KeyAgreement implementation for ECDH.
+ */
+public final class NativeECDHKeyAgreement extends KeyAgreementSpi {
+
+    private static final NativeCrypto nativeCrypto = NativeCrypto.getNativeCrypto();
+    private static final String nativeCryptTrace = GetPropertyAction.privilegedGetProperty("jdk.nativeCryptoTrace");
+
+    /* false if OPENSSL_NO_EC2M is defined, true otherwise */
+    private static final boolean nativeGF2m = nativeCrypto.ECNativeGF2m();
+
+    /* stores whether a curve is supported by OpenSSL (true) or not (false) */
+    private static final Map<String, Boolean> curveSupported = new ConcurrentHashMap<>();
+
+    /* private key, if initialized */
+    private ECPrivateKeyImpl privateKey;
+
+    /* public key, non-null between doPhase() & generateSecret() only */
+    private ECPublicKeyImpl publicKey;
+
+    /* the type of EC curve */
+    private String curve;
+
+    /* length of the secret to be derived */
+    private int secretLen;
+
+    /* the java implementation, initialized if needed */
+    private ECDHKeyAgreement javaImplementation;
+
+    /**
+     * Constructs a new NativeECDHKeyAgreement.
+     */
+    public NativeECDHKeyAgreement() {
+    }
+
+    @Override
+    protected void engineInit(Key key, SecureRandom random)
+            throws InvalidKeyException {
+        if (!(key instanceof PrivateKey)) {
+            throw new InvalidKeyException
+                ("Key must be an instance of PrivateKey");
+        }
+        /* attempt to translate the key if it is not an ECKey */
+        this.privateKey = (ECPrivateKeyImpl) ECKeyFactory.toECKey(key);
+        this.publicKey = null;
+
+        ECParameterSpec params = this.privateKey.getParams();
+        if (params instanceof NamedCurve) {
+            this.curve = ((NamedCurve) params).getName();
+        } else {
+            /* use the OID */
+            try {
+                AlgorithmParameters algParams = AlgorithmParameters.getInstance("EC");
+                algParams.init(this.privateKey.getParams());
+                this.curve = algParams.getParameterSpec(ECGenParameterSpec.class).getName();
+            } catch (NoSuchAlgorithmException e) {
+                /* should not happen */
+            } catch (InvalidParameterSpecException e) {
+                System.out.println(e.toString());
+            }
+        }
+
+        if ((!nativeGF2m) && this.privateKey.isECFieldF2m()) {
+            if (curveSupported.containsKey("EC2m")) {
+                this.useJavaImplementation(key, random, null);
+            } else {
+                curveSupported.put("EC2m", Boolean.FALSE);
+                this.useJavaImplementation(key, random, "EC2m");
+            }
```

Comment:
Elsewhere, the third argument is `this.curve`; I suggest the same here which allows for simplification:
```
            curveSupported.putIfAbsent("EC2m", Boolean.FALSE);
            this.useJavaImplementation(key, random, "EC2m");
```

Reply:
Perhaps that tracing should be extracted from `useJavaImplementation()` and called separately.
At the very least, we should have a comment explaining that strategy.

Reply:
done